PredicateKit/CoreData/NSManagedObjectContextExtensions.swift

```diff
@@ -370,6 +370,28 @@ public struct FetchRequest<Entity: NSManagedObject> {
     let request: NSFetchRequest<Entity> = requestBuilder.makeRequest(from: self)
     return try context.count(for: request)
   }
+  
+   /// An NSFetchedResultsController initialized with the fetch request.
+   ///
+   /// - Parameters:
+   ///    - sectionNameKeyPath - keypath on resulting objects that returns the section name. This will be used to pre-compute the section information. Defaults to `nil`.
+   ///    - cacheName - Section info is cached persistently to a private file under this name. Cached sections are checked to see if the time stamp matches the store, but not if you have illegally mutated the readonly fetch request, predicate, or sort descriptor. Defaults to `nil`.
+   ///
+   /// - Returns: A fetchedResultsController with objects of type `Entity` matching the criteria specified by the fetch request.
+   ///
+   /// ## Example
+   ///
+   ///      let notes: NSFetchedResultsController<Note> = managedObjectContext
+   ///        .fetch(where: (\Note.text).contains("Hello, World!"))
+   ///        .sorted(by: \.creationDate, .descending)
+   ///        .fetchedResultsController()
+   ///
+   public func fetchedResultsController(sectionNameKeyPath: String? = nil, cacheName: String? = nil) -> NSFetchedResultsController<Entity> {
+     let request: NSFetchRequest<Entity> = requestBuilder.makeRequest(from: self)
+     request.resultType = .managedObjectResultType
+     return NSFetchedResultsController(fetchRequest: request, managedObjectContext: self.context,
+                                       sectionNameKeyPath: sectionNameKeyPath, cacheName: cacheName)
+   }
```

Comment:
What do you think if we used a `KeyPath` for the parameter `sectionNameKeyPath`. This will add type safety and guarantee that we use only valid and comparable keys. Something like the following:

```suggestion
  public func fetchedResultsController<T: Comparable & Primitive>(
    sectionNameKeyPath: KeyPath<Entity, T>? = nil,
    cacheName: String? = nil
  ) -> NSFetchedResultsController<Entity> {
    let request: NSFetchRequest<Entity> = requestBuilder.makeRequest(from: self)
    request.resultType = .managedObjectResultType
    return NSFetchedResultsController(
      fetchRequest: request,
      managedObjectContext: context,
      sectionNameKeyPath: sectionNameKeyPath?.stringValue,
      cacheName: cacheName
    )
  }
```

The usage will then be

```swift
let notes: NSFetchedResultsController<Note> = managedObjectContext
  .fetch(where: (\Note.text).contains("Hello, World!"))
  .sorted(by: \.creationDate, .descending)
  .fetchedResultsController(sectionNameKeyPath: \.creationDate)
```

**vs**

```swift
let notes: NSFetchedResultsController<Note> = managedObjectContext
  .fetch(where: (\Note.text).contains("Hello, World!"))
  .sorted(by: \.creationDate, .descending)
  .fetchedResultsController(sectionNameKeyPath: "creationDate")
```

This will require [this extension](https://github.com/ftchirou/PredicateKit/blob/main/PredicateKit/CoreData/NSFetchRequestBuilder.swift#L301) to be made internal.

Reply:
You beat me to it! This was next on my list.

Reply:
There are a couple of issues with using KeyPath<Entity, T> for sectionNameKeyPath:
1. If sectionNameKeyPath is nil, T cannot be inferred and the function fails at compile time. In actuality, this doesn't really matter, because...

2. This function needs to be able to accept various types of KeyPaths (ReferenceWritableKeyPath for @NSManaged properties, KeyPath for computed properties, etc.). I think this would be a good place for one of the type-erased KeyPath variants, either PartialKeyPath<Entity> or AnyKeyPath. I lean toward PartialKeyPath, since it enforces Root, but am open to other arguments or ideas.

Thoughts?

Reply:
> If sectionNameKeyPath is nil, T cannot be inferred and the function fails at compile time.

Indeed, you're correct.

> I think this would be a good place for one of the type-erased KeyPath variants, either PartialKeyPath or AnyKeyPath. I lean toward PartialKeyPath, since it enforces Root

The only concern I would have using `PartialKeyPath` is that it enforces `Root` but does not enforce `Value`. That means users could in theory use any key-path for `sectionNameKeyPath`. For example, with the following classes,

```swift
class User: NSManagedObject {
  @NSManaged var name: String
  @NSManaged var billingInfo: BillingInfo
}

class BillingInfo: NSManagedObject {
  @NSManaged var accountType: String
  @NSManaged var purchases: [Double]
}
```

it would be valid to write:

```swift
let controller = managedObjectContext
  .fetch(where: (\User.name).contains("Joe"))
  .sorted(by: \User.billingInfo.accountType)
  .fetchedResultsController(sectionNameKeyPath: \User.billingInfo)
```

❌ However `billingInfo` is not a primitive nor comparable.

> but am open to other arguments or ideas.

I think we can 2 functions, one with a non-optional `sectionNameKeyPath: KeyPath<Entity, T>` and another without.

```swift
public func fetchedResultsController<T: Comparable & Primitive>(
  sectionNameKeyPath: KeyPath<Entity, T>,
  cacheName: String? = nil
) -> NSFetchedResultsController<Entity> {
  fetchedResultsController(sectionNameKeyPath: sectionNameKeyPath.stringValue, cacheName: cacheName)
}

public func fetchedResultsController(cacheName: String? = nil) -> NSFetchedResultsController<Entity> {
  fetchedResultsController(sectionNameKeyPath: nil, cacheName: cacheName)
}

private func fetchedResultsController(
  sectionNameKeyPath: String?,
  cacheName: String? = nil
) -> NSFetchedResultsController<Entity> {
  let request: NSFetchRequest<Entity> = requestBuilder.makeRequest(from: self)
  request.resultType = .managedObjectResultType
  return NSFetchedResultsController(
    fetchRequest: request,
    managedObjectContext: context,
    sectionNameKeyPath: sectionNameKeyPath,
    cacheName: cacheName
  )
}
```

It's slightly more involved but we gain type-safety 😄. What do you think?

Reply:
That works for me.